src/dict.js

```diff
@@ -69,68 +74,64 @@ Sk.builtin.dict.tp$call = function(args, kw) {
 Sk.abstr.setUpInheritance("dict", Sk.builtin.dict, Sk.builtin.object);
 Sk.abstr.markUnhashable(Sk.builtin.dict);
 
-var kf = Sk.builtin.hash;
-
-Sk.builtin.dict.prototype.key$lookup = function (bucket, key) {
-    var item;
-    var eq;
-    var i;
-
-    // Fast path: We spend a *lot* of time looking up strings
-    // in dictionaries. (Every attribute access, for starters.)
-    if (key.ob$type === Sk.builtin.str) {
-        for (i = 0; i < bucket.items.length; i++) {
-            item = bucket.items[i];
-            if (item.lhs.ob$type === Sk.builtin.str && item.lhs.v === key.v) {
-                return item;
-            }
-        }
-        return null;
+function kf(key) {
+    let key_hash;
+    if (key.$savedKeyHash_) {
+        return key.$savedKeyHash_;
     }
-
-    for (i = 0; i < bucket.items.length; i++) {
-        item = bucket.items[i];
-        eq = Sk.misceval.richCompareBool(item.lhs, key, "Eq");
-        if (eq) {
-            return item;
-        }
+    if (Sk.builtin.checkString(key)) {
+        key_hash = "_" + key.$jsstr();
+        key.$savedKeyHash_ = key_hash;
+        return key_hash;
+    } else if (typeof key === "string") {
+        return "_" + key;
     }
-
-    return null;
+    key_hash = "#_" + Sk.builtin.hash(key).v;
+    key.$savedKeyHash_ = key_hash; // this is a base key hash
+    return key_hash;
 };
 
-Sk.builtin.dict.prototype.key$pop = function (bucket, key) {
-    var item;
-    var eq;
-    var i;
+Sk.builtin.dict.prototype.sk$asarray = function () {
+    const all_keys = [];
+    const entries = this.entries;
+    for (let hash in entries) {
+        all_keys.push(entries[hash].lhs);
+    }
+    return all_keys;
+};
 
-    for (i = 0; i < bucket.items.length; i++) {
-        item = bucket.items[i];
-        eq = Sk.misceval.richCompareBool(item.lhs, key, "Eq");
-        if (eq) {
-            bucket.items.splice(i, 1);
-            this.size -= 1;
-            return item;
+Sk.builtin.dict.prototype.get$item_from_bucket = function (key, base_hash) {
+    const bucket = this.buckets[base_hash];
+    let stored_key, item;
+    if (bucket === undefined) {
+        return;
+    }
+    for (let i=0; i<bucket.length; i++) {
+        item = bucket[i];
+        if (item === undefined) {
+            continue;
+        }
+        stored_key = item.lhs;
+        if (stored_key === key || Sk.misceval.richCompareBool(key, stored_key, "Eq")) {
+            return this.entries["#" + i + base_hash.slice(1,)];
```

Comment:
This method of construction confused me a little - it wasn't obvious that `base_hash.slice(1,)` will *always* start with an underscore, so I assumed there might be ambiguity here.
 
I think this will be resolved by documenting more about how this works in a code comment.

Reply:
realised I could change this specific line to `return item`
but generally i've added some additional comments for clarity